src/js/autocomplete.controller.js

```diff
@@ -1,542 +1,559 @@
-            self.element.input.blur();
-            self.element.input.value = null;
-            handleSearchText();
-            self.element.input.focus();
-        }
-
+  'use strict';
+
+  var MaterialAutocompleteCntrl = function ($scope, $element, $q, $timeout, $filter) {
+    var self = this;
+
+    $timeout(function () {
+      self.parentForm = $scope.parentForm;
+      self.immediateParentForm = $scope.immediateParentForm;
+    }, 20);
+
+
+    /**
+     * Common Keyboard actions and their associated keycode.
+     */
+    var KEY_CODE = {
+      COMMA: 188,
+      SEMICOLON: 186,
+      ENTER: 13,
+      ESCAPE: 27,
+      SPACE: 32,
+      PAGE_UP: 33,
+      PAGE_DOWN: 34,
+      END: 35,
+      HOME: 36,
+      LEFT_ARROW: 37,
+      UP_ARROW: 38,
+      RIGHT_ARROW: 39,
+      DOWN_ARROW: 40,
+      TAB: 9,
+      BACKSPACE: 8,
+      DELETE: 46
+    };
 
-        /**
-         * Sets up any watchers used by autocomplete
-         */
-        function configureWatchers() {
-            // $attrs.$observe('disabled', function (value) { ctrl.isDisabled = $mdUtil.parseAttributeBoolean(value, false); });
-            // $attrs.$observe('required', function (value) { ctrl.isRequired = $mdUtil.parseAttributeBoolean(value, false); });
-            // $attrs.$observe('readonly', function (value) { ctrl.isReadonly = $mdUtil.parseAttributeBoolean(value, false); });
-
-            $scope.$watch(angular.bind(self, function () {
-                return self.searchText;
-            }), handleSearchText);
-            $scope.$watch(angular.bind(self, function () {
-                return self.selectedItem;
-            }), selectedItemChange);
-
-            // angular.element($window).on('resize', debouncedOnResize);
-            //
-            // $scope.$on('$destroy', cleanup);
-        }
+    var noBlur = false,
+      hasFocus = false;
+
+    self.required = !!self.required;
+    self.isInputFocus = false;
+    self.isInputBlur = false;
+    self.clearButton = false;
+    self.loading = false;
+    self.index = -1;
+    self.isDisabled = null;
+    self.isRequired = null;
+    self.hasNotFound = false;
+    self.hidden = true;
+    if (!self.searchText) {
+      self.searchText = '';
+    }
+
+    self.init = function () {
+      configureWatchers();
+      self.element = {
+        input: $element.find('input')[0],
+        scroller: $element.find('ul')[0],
+        scrollContainer: $element[0].querySelector('#acDropdown'),
+        label: $element.find('label')[0]
+      };
+      self.element.li = self.element.scroller.getElementsByTagName('li');
+      self.element.$ = getAngularElements(self.element);
+
+      self.setInputName();
+
+      if (self.itemList && typeof self.itemList[0] === 'string') {
+        self.itemList = convertArrayToObject(self.itemList);
+      }
+
+      if (self.itemList && typeof self.itemList[0] === 'object') {
+        self.itemListCopy = angular.copy(self.itemList);
```

Comment:
please explain this code

Reply:
I am storing contents of the `self.itemList` into another variable, because the contents of `self.itemList` change when a filter is applied to it.
For example, if you search `product`, all other values go away from the list. On pressing the cross button, you need the values back, right? That's where `self.itemListCopy` comes handy.